lib/matplotlib/axes/_base.py

```diff
@@ -229,24 +228,79 @@ def _xy_from_xy(self, x, y):
             y = y[:, np.newaxis]
         return x, y
 
+
+    def _getdefaults(self, *kwargs):
+        """
+        Only advance the cycler if the cycler has information that
+        is not specified in any of the supplied tuple of dicts
+
+        Returns a dictionary or None.
+
+        """
+        if any([all([kw.get(k, None) is None for kw in kwargs])
+               for k in self._prop_keys]):
+            default_dict = six.next(self.prop_cycler)
+        else:
+            default_dict = None
+        return default_dict
+
+
+    def _setdefaults(self, default_dict, *kwargs):
+        """
+        Given a defaults dictionary (or None), and a other dictionaries,
+        update the other dictionaries with information in defaults if
+        none of the other dictionaries contains that information.
+
+        """
+        if default_dict is None:
+            return
+        for k in default_dict:
+            if all([kw.get(k, None) is None for kw in kwargs]):
+                for kw in kwargs:
+                    kw[k] = default_dict[k]
+
+
     def _makeline(self, x, y, kw, kwargs):
         kw = kw.copy()  # Don't modify the original kw.
         kwargs = kwargs.copy()
-        if kw.get('color', None) is None and kwargs.get('color', None) is None:
-            kwargs['color'] = kw['color'] = six.next(self.color_cycle)
-            # (can't use setdefault because it always evaluates
-            # its second argument)
-        seg = mlines.Line2D(x, y,
-                            **kw
-                            )
+        default_dict = self._getdefaults(kw, kwargs)
+        self._setdefaults(default_dict, kw, kwargs)
+        seg = mlines.Line2D(x, y, **kw)
         self.set_lineprops(seg, **kwargs)
         return seg
 
+
     def _makefill(self, x, y, kw, kwargs):
-        try:
-            facecolor = kw['color']
-        except KeyError:
-            facecolor = six.next(self.color_cycle)
+        kw = kw.copy()  # Don't modify the original kw.
+        kwargs = kwargs.copy()
+
+        # Only using the first dictionary to use as basis
+        # for getting defaults for back-compat reasons.
+        # Doing it with both seems to mess things up in
+        # various places (probably due to logic bugs elsewhere).
+        default_dict = self._getdefaults(kw)
+        self._setdefaults(default_dict, kw)
+
+        # Looks like we don't want "color" to be interpreted to
+        # mean both facecolor and edgecolor for some reason.
+        # So the "kw" dictionary is thrown out, and only its
+        # 'color' value is kept and translated as a 'facecolor'.
+        # This design should probably be revisited as it increases
+        # complexity.
+        facecolor = kw.get('color', None)
```

Comment:
All of this extra code is just to get behavior similar to before, that is that for things like bar graphs, a color is really only a facecolor and not also the edgecolor. If the user supplied a facecolor, then that would be handled the same as always through `kwargs` dictionary.